usr/src/common/cmd/ksh93/src/cmd/ksh93/sh.1
.\"
.\" David Korn
.\" AT&T Bell Laboratories
.\"
.\" @(#)sh.1 (research!dgk) 12/28/93
.\"
.nr Z 0 \" set to 1 when command name is ksh rather than sh
.nr Y 0 \" set to 1 for R&D UNIX
.if \nY=1 .nr Z 1
.ds OK [\|
.ds CK \|]
.ds ' \s+4\v@.3m@\'\v@-.3m@\s-4
.ds ` \s+4\v@.3m@\`\v@-.3m@\s-4
.if \nZ=0 .TH SH 1
.if \nZ=1 .TH KSH 1 "User Environment Utilities" "RDS Standard"
.SH NAME
.if \nZ=0 sh, rsh \- shell, the
.if \nZ=1 ksh, rksh \- KornShell, a
standard/restricted command and programming language
.SH SYNOPSIS
.if \nZ=0 .B sh
.if \nZ=1 .B ksh
[
.B \(+-abcefhikmnoprstuvxCD
] [
.B \-I
file ] [
.B \(+-o
option ] .\|.\|. [
.B \-
] [ arg .\|.\|. ]
.br
.if \nZ=0 .B rsh
.if \nZ=1 .B rksh
[
.B \(+-abcefhikmnoprstuvxCD
] [
.B \-I
file ] [
.B \(+-o
option ] .\|.\|.  [
.B \-
] [ arg .\|.\|. ]
.SH DESCRIPTION
.if \nZ=0 .I Sh\^
.if \nZ=1 .I Ksh\^
is a command and programming language
that executes commands read from a terminal
or a file.
.if \nZ=0 .I Rsh\^
.if \nZ=1 .I Rksh\^
is a restricted version of the 
.if \nZ=0 standard 
command interpreter
.if \nZ=0 .IR sh ;
.if \nZ=1 .IR ksh ;
it is used to set up login names and execution environments whose
capabilities are more controlled than those of the standard shell.
See
.I Invocation\^
below
for the meaning of arguments to the shell.
.SS Definitions.
A
.I metacharacter\^
is one of the following characters:
.PP
.RS
\f3;   &   (   )   \(bv   <   >   new-line   space   tab\fP
.RE
.PP
A
.I blank\^
is a
.B tab
or a
.BR space .
An
.I identifier\^
is a sequence of letters, digits, or underscores
starting with a letter or underscore.
Identifiers are used as components of
.I variable\^
names.
A
.I vname\^
is a sequence of one or more identifiers
separated by a \fB\s+2.\s-2\fP and optionally preceded
by a \fB\s+2.\s-2\fP.
Vnames are used as function and variable names.
A
.I word\^
is a sequence of
.I characters\^
from the character set defined by the current locale,
excluding non-quoted
.IR metacharacters .
.PP
A 
.I command\^
is a sequence of characters in the syntax
of the shell language.
The shell reads each command and
carries out the desired action either directly or by invoking
separate utilities.
A built-in command is a command that is carried out by the
shell itself without creating a separate process.
Some commands are built-in purely for convenience
and are not documented here.
Built-ins that cause
side effects in the shell environment and 
built-ins that are found before performing a
path search (see
.I Execution\^
below)
are documented here.
For historical reasons, some of
these built-ins behave differently than
other built-ins and are called
.IR "special built-ins" .
.SS Commands.
A
.I simple-command\^
is a list of variable assignments
(see
.I Variable Assignments\^
below)
or a sequence of
.I blank\^
separated words
which may be preceded by a list of variable assignments
(see
.I Environment\^
below).
The first word specifies the name of the command to
be executed.
Except as specified below,
the remaining words are passed as arguments
to the invoked command.
The command name is passed as argument 0
(see
.IR exec (2)).
The
.I value\^
of a simple-command is its exit status; 0-255
if it terminates normally; 256+\f2signum\^\fP if
it terminates abnormally (the name of the signal corresponding
to the exit status can be
obtained via the
.B \-l
option of the
.B kill\^
built-in utility).
.PP
A
.I pipeline\^
is a sequence of one or more
.I commands\^
separated by
.BR \(bv .
The standard output of each command but the last
is connected by a
.IR pipe (2)
to the standard input of the next command.
Each command,
except possibly the last,
is run as a separate process;
the shell waits for the last command to terminate.
The exit status of a pipeline is the exit
status of the last command.
Each pipeline can be preceded by the
.I "reserved word"
.B !
which causes the exit status of the pipeline to become
0 if the exit status of the last command is non-zero, and
1 if the exit status of the last command is 0.
.PP
A
.I list\^
is a sequence of one or more
pipelines
separated by
.BR ; ,
.BR & ,
.BR \(bv& ,
.BR && ,
or
.BR \(bv\|\(bv ,
and optionally terminated by
.BR ; ,
.BR & ,
or
.BR \(bv& .
Of these five symbols,
.BR ; ,
.BR & ,
and
.B \(bv&
have equal precedence,
which is lower than that of
.B &&
and
.BR \(bv\|\(bv .
The symbols
.B &&
and
.B \(bv\|\(bv
also have equal precedence.
A semicolon
.RB ( ; )
causes sequential execution of the preceding pipeline; an ampersand
.RB ( & )
causes asynchronous execution of the preceding pipeline (i.e., the shell does
.I not\^
wait for that pipeline to finish).
The symbol
.B \(bv&
causes asynchronous execution of the preceding pipeline
with a two-way pipe established to the parent shell;
the standard input and output of the spawned pipeline
can be written to and read from by the parent shell
by applying
the redirection operators
.B <&
and
.B >&
with arg
.B p
to commands and by using
.B \-p
option of
the built-in commands
.B read
and
.B print
described later.
The symbol
.B &&
.RB (\| \(bv\|\(bv \^)
causes the
.I list\^
following it to be executed only if the preceding
pipeline
returns a zero (non-zero) value.
One or more new-lines may appear in a
.I list\^
instead of a semicolon,
to delimit a command.
.PP
A
.I command\^
is either a simple-command
or one of the following.
Unless otherwise stated,
the value returned by a command is that of the
last simple-command executed in the command.
.TP
\f3for\fP \f2vname\^\fP \*(OK \f3in\fP \f2word\^\fP .\|.\|. \*(CK \f3;do\fP \f2list\^\fP \f3;done\fP
Each time a
.B for
command is executed,
.I vname\^
is set to the next
.I word\^
taken from the
.B in
.I word\^
list.
If
.BI in " word\^"
\&.\|.\|.
is omitted, then
the
.B for
command executes the \f3do\fP \f2list\^\fP once for each positional parameter
that is set starting from
.B 1
(see
.I "Parameter Expansion\^"
below).
Execution ends when there are no more words in the list.
.TP
\f3for ((\fP \*(OK\f2expr1\^\fP\*(CK \f3;\fP \*(OK\f2expr2\^\fP\*(CK \f3;\fP \*(OK\f2expr3\^\fP\*(CK \f3))\fP \f3;do\fP \f2list\^\fP \f3;done\fP
The arithmetic expression
.I expr1
is evaluated first
(see
.I "Arithmetic evaluation"
below).
The arithmetic expression
.I expr2
is repeatedly evaluated until it evaluates to zero and when non-zero,
.I list
is executed and the arithmetic expression
.I expr3
evaluated.
If any expression
is omitted, then it behaves as if it evaluated to 1.
.TP
\f3select\fP \f2vname\^\fP \*(OK \f3in\fP \f2word\^\fP .\|.\|. \*(CK \f3;do\fP \f2list\^\fP \f3;done\fP
A
.B select
command prints on standard error (file descriptor 2) the set of
.IR word s,
each preceded by a number.
If
.BI in " word\^"
\&.\|.\|.
is omitted, then
the
positional parameters starting from
.B 1
are used instead
(see
.I "Parameter Expansion\^"
below).
The
.SM
.B PS3
prompt is printed
and a line is read from the standard input.
If this line consists of the number
of one of the listed
.IR word s,
then the value of the variable
.I vname\^
is set to the
.I word\^
corresponding to this number.
If this line is empty, the selection list is
printed again.
Otherwise the value of the variable
.I vname\^
is set to
.IR null .
The contents of the line read from standard input is
saved in
the variable
.SM
.BR REPLY .
The
.I list\^
is executed for each selection until a
.B break\^
or
.I end-of-file\^
is encountered.
If the
.SM
.B REPLY
variable is set to
.I null\^
by the execution of
.IR list ,
then the selection list is printed before
displaying the
.SM
.B PS3
prompt for the next selection.
.TP
\f3case\fP \f2word\^\fP \f3in\fP \*(OK \*(OK\f3(\fP\*(CK\f2pattern\^\fP \*(OK \(bv \f2pattern\^\fP \*(CK .\|.\|. \f3)\fP \f2list\^\fP \f3;;\fP \*(CK .\|.\|. \f3esac\fP
A
.B case
command executes the
.I list\^
associated with the first
.I pattern\^
that matches
.IR word .
The form of the patterns is
the same as that used for
file-name generation (see
.I "File Name Generation\^"
below).
The
.B ;;
operator causes execution of
.B case
to terminate.
If
.B ;&
is used in place of
.B ;;
the subsequent list, if any, is executed.
.TP
\f3if\fP \f2list\^\fP \f3;then\fP \f2list\^\fP \*(OK \
\f3elif\fP \f2list\^\fP \f3;then\fP \f2list\^\fP \*(CK .\|.\|. \
\*(OK \f3;else\fP \f2list\^\fP \*(CK \f3;f\&i\fP
The
.I list\^
following \f3if\fP is executed and,
if it
returns a zero exit status, the
.I list\^
following
the first
.B then
is executed.
Otherwise, the
.I list\^
following \f3elif\fP
is executed and, if its value is zero,
the
.I list\^
following
the next
.B then
is executed.
Failing each successive
.B elif
.IR list\^ ,
the
.B else
.I list\^
is executed.
If the
.B if
.I list\^
has non-zero exit status
and there is no
.B else
.IR list ,
then the
.B if
command returns a zero exit status.
.TP
.PD 0
\f3while\fP \f2list\^\fP \f3;do\fP \f2list\^\fP \f3;done\fP
.TP
\f3until\fP \f2list\^\fP \f3;do\fP \f2list\^\fP \f3;done\fP
.PD
A
.B while
command repeatedly executes the
.B while
.I list\^
and, if the exit status of the last command in the list is zero, executes
the
.B do
.IR list ;
otherwise the loop terminates.
If no commands in the
.B do
.I list\^
are executed, then the
.B while
command returns a zero exit status;
.B until
may be used in place of
.B while
to negate
the loop termination test.
.TP
\f3((\fP\f2expression\^\fP\f3))\fP
.br
The
.I expression\^
is evaluated using the rules for arithmetic evaluation described below.
If the value of the arithmetic expression is non-zero, the exit
status is 0, otherwise the exit status is 1.
.TP
\f3(\fP\f2list\^\fP\f3)\fP
.br
Execute
.I list\^
in a separate environment.
Note, that if two adjacent open parentheses are
needed for nesting, a space must be inserted to avoid
evaluation as an arithmetic command as described above.
.TP
\f3{ \fP\f2list\^\fP\f3;}\fP
.br
.I list\^
is simply executed.
Note that unlike the metacharacters
.B (
and
.BR ) ,
.B {
and
.B }
are
.IR "reserved word" s
and must occur
at the beginning of a line or after a
.B ;
in order to be recognized.
.TP
\f3[[\fP\f2 expression \^\fP\f3]]\fP
.br
Evaluates
.I expression\^
and returns a zero exit status when
.I expression\^
is true.
See
.I "Conditional Expressions\^"
below, for a description of
.IR expression .
.TP
.PD 0
\f3function\fP \f2varname\^\fP \f3{\fP \f2list\^\fP \f3;}\fP
.TP
\f2varname\^\fP \f3() {\fP \f2list\^\fP \f3;}\fP
.PD
Define a function which is referenced by
.IR varname .
A function whose
.I varname\^
contains a
.B \s+2.\s-2
is called a discipline function and the portion
of the
.I varname\^
preceding the last
.B \s+2.\s-2
must refer to an existing variable. 
The body of the function is the
.I list\^
of commands between
.B {
and
.BR } .
A function defined with the \f3function\fP \f2varname\^\fP
syntax can also be used as an argument to the \f3.\fP
special built-in command to get the equivalent behavior
as if the \f2varname\^\fP\f3()\fP syntax were used to define it.
(See
.I Functions\^
below.)
.TP
\f3time\fP \*(OK \f2pipeline\^\fP \*(CK
.br
If \f2pipeline\^\fP is omitted the user and system time for
the current shell and completed child processes is printed
on standard error.
Otherwise,
.I pipeline\^
is executed and the elapsed time as well as
the user and system time are printed on standard error.
.PP
The following reserved words
are recognized as reserved only when they serve as the first word of a command

usr/src/common/cmd/OSR/csh/sh.c
#ident	"@(#)OSRcmds:csh/sh.c	1.1"
/*
 *	@(#) sh.c 26.2 96/03/14 
 *
 *	      UNIX is a registered trademark of AT&T
 *		Portions Copyright 1976-1989 AT&T
 *	Portions Copyright 1980-1989 Microsoft Corporation
 *   Portions Copyright 1983-1996 The Santa Cruz Operation, Inc
 *		      All Rights Reserved
 */
/*	Copyright (c) 1984, 1986, 1987, 1988 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

/*	Copyright (c) 1987, 1988 Microsoft Corporation	*/
/*	  All Rights Reserved	*/

/*	This Module contains Proprietary Information of Microsoft  */
/*	Corporation and should be treated as Confidential.	   */

/* #ident	"@(#)csh:sh.c	1.4" */
#pragma comment(exestr, "@(#) sh.c 26.2 96/03/14 ")

/*
 *	@(#) sh.c 1.2 88/03/30 csh:sh.c
 */
/* Copyright (c) 1980 Regents of the University of California */
#include "sh.h"
extern	char	*sbrk();	/* M000 */

/*
 * C Shell
 *
 * Bill Joy, UC Berkeley
 * October, 1978
 *
 *	MODIFICATION HISTORY
 *	M000	18 Oct 83	andyp
 *	- Took out "login" command, now conditional on LOGIN.  XENIX 3.0
 *	  does not support this as a regular command.
 *	- Removed (v7) assumption that "" == "." for pathnames.
 *	- Removed dependency on pseudo-symbol "end" produced by ld, and
 *	  assumptions about location of stack relative to end of allocated
 *	  data.  This change necessary for fixed-stack machines.
 *	- Changed illegal old-style C-common declarations to externs.
 *	M001	26 Feb 84	andyp
 *	Thanks to Greg Guthrie at Bell Labs, Naperville and net.bugs
 *	- Fix for escaped quotes in quoted strings.
 *	M002	15 May 84	barrys
 *	- Signals are now turned on before executing a newgrp. Before
 *	  this resulted in a user doing a newgrp and then not able
 *	  to perform a quit.
 *	- Changed the minimum number of arguments to newgrp from 1
 *	  to 0.
 *	M003	25 July 84	davep
 *	- source /etc/cshrc for users that have csh as their login shell
 *	  Allows the adminstrator to have default things set up for every user
 *	  Note: This is done before .cshrc or .login so that the user can
 *		override it if necessary.
 *	S004	4 Apr 1989	sco!hoeshuen
 *	- If interactive, set shell variable "cdspell".  If set,
 *	  cd spelling correction occurs.
 *	L005	17 Feb 1994	scol!gregw
 *	- Renamed calloc to Calloc to avoid possible conflict with libc.
 *	L006	27 Nov 1995	scol!ianw
 *	- Changed the type of some automatic variables to static to ensure that
 *	  on return from setjmp they still have the values they had when
 *	  longjmp was called.
 *	L007	14 Mar 1996	scol!ianw
 *	- Fixed a bug introduced by L006 (when a .login file uses the source
 *	  built-in command the user is logged out) by changing the static
 *	  variables to volatile, necessary as srcunit() can be called
 *	  recursively (srcunit() calls process() which can lead to dosource()
 *	  being called which itself calls srcunit()).
 */

char	*pathlist[] =	{ SRCHPATH, 0 };
char	*End;			/* M000 */

main(c, av)
	int c;
	char **av;
{
	register char **v, *cp;
	int nofile = 0;
	volatile int reenter = 0;				/* L006 L007 */
	bool nverbose = 0, nexececho = 0, quitit = 0, fast = 0, prompt = 1;
	char *hp;

	End = sbrk(0);			/* M000 */
	settimes();			/* Immed. estab. timing base */
	hp = getenv("HOME");
	v = av;
	if (eq(v[0], "a.out"))		/* A.out's are quittable */
		quitit = 1;
	uid = getuid();
#ifdef V6
	loginsh = eq(*v, "-");		/* To do .login/.logout */
#else
	loginsh = **v == '-';
#endif
	if (loginsh)
		time(&chktim);

	/*
	 * Move the descriptors to safe places.
	 * The variable didfds is 0 while we have only FSH* to work with.
	 * When didfds is true, we have 0,1,2 and prefer to use these.
	 */
	initdesc();

	/*
	 * Initialize the shell variables.
	 * ARGV and PROMPT are initialized later.
	 * STATUS is also munged in several places.
	 * CHILD is munged when forking/waiting
	 */

	set("status", "0");

	/* Default history characters */
	HIST = '!'; HISTSUB = '^';

	if (hp == 0)
		fast++;			/* No home -> can't read scripts */
	else
		set("home", hp);
	if (uid == 0)
		pathlist[0] = "/etc";
	set1("path", saveblk(pathlist), &shvhed);
	/*
	 * Re-initialize path if set in environment
	 */
	cp = getenv("PATH");
	if (cp != 0) {
		importpath(cp);
	}

	set("shell", SHELLPATH);

	doldol = putn(getpid());		/* For $$ */
	shtemp = strspl("/tmp/sh", doldol);	/* For << */

	/*
	 * Record the interrupt states from the parent process.
	 * If the parent is non-interruptible our hand must be forced
	 * or we (and our children) won't be either.
	 * Our children inherit termination from our parent.
	 * We catch it only if we are the login shell.
	 */
	parintr = signal(SIGINT, SIG_IGN);	/* parents interruptibility */
	signal(SIGINT, parintr);			/* ... restore */
	parterm = signal(SIGTERM, SIG_IGN);	/* parents terminability */
	signal(SIGTERM, parterm);			/* ... restore */

	/*
	 * Process the arguments.
	 *
	 * Note that processing of -v/-x is actually delayed till after
	 * script processing.
	 *
	 * We set the first character of our name to be '-' if we are
	 * a shell running interruptible commands.  Many programs which
	 * examine ps'es use this to filter such shells out.
	 */
	c--, v++;
	while (c > 0 && (cp = v[0])[0] == '-') {
		do switch (*cp++) {

		case 0:			/* -	Interruptible, no prompt */
			prompt = 0;
			**av = '-';
			nofile++;
			break;

		case 'c':		/* -c	Command input from arg */
			if (c == 1)
				exit(0);
			c--, v++;
			arginp = v[0];
			prompt = 0;
			nofile++;
			break;

		case 'e':		/* -e	Exit on any error */
			exiterr++;
			break;

		case 'f':		/* -f	Fast start */
			fast++;
			break;

		case 'i':		/* -i	Interactive, even if !intty */
			intact++;
			**av = '-';
			nofile++;
			break;

		case 'n':		/* -n	Don't execute */
			noexec++;
			break;

		case 'q':		/* -q	(Undoc'd) ... die on quit */
			quitit = 1;
			break;

		case 's':		/* -s	Read from std input */
			nofile++;
			if (isatty(SHIN))
				**v = '-';
			break;

		case 't':		/* -t	Read one line from input */
			onelflg = 2;
			if (isatty(SHIN))
				**v = '-';
			prompt = 0;
			nofile++;
			break;

		case 'v':		/* -v	Echo hist expanded input */
			nverbose = 1;			/* ... later */
			break;

		case 'x':		/* -x	Echo just before execution */
			nexececho = 1;			/* ... later */
			break;

		case 'V':		/* -V	Echo hist expanded input */
			setNS("verbose");		/* NOW! */
			break;

		case 'X':		/* -X	Echo just before execution */
			setNS("echo");			/* NOW! */
			break;

		} while (*cp);
		v++, c--;
	}

	if (quitit)			/* With all due haste, for debugging */
		signal(SIGQUIT, SIG_DFL);

	/*
	 * Unless prevented by -, -c, -i, -s, or -t, if there
	 * are remaining arguments the first of them is the name
	 * of a shell file from which to read commands.
	 */
	if (nofile == 0 && c > 0) {
		nofile = open(v[0], 0);
		if (nofile < 0) {
			child++;		/* So this ... */
			Perror(v[0]);		/* ... doesn't return */
		}
		file = v[0];
		SHIN = dmove(nofile, FSHIN);	/* Replace FSHIN */
		prompt = 0;
		c--, v++;
	}

	/*
	 * Consider input a tty if it really is or we are interactive.
	 */
	intty = intact || isatty(SHIN);
#ifdef TELL
	settell();
#endif

	/* Begin SCO_BASE S004...
	 * If we are reading from a tty, then default is cd spelling
	 * correction on.  Note that if we're not reading from a tty,
	 * then no spelling correction occurs reguardless of "cdspell".
	 */
	if (intty)
		setNS("cdspell");
	/* End S004...

	/*
	 * Commands are interruptible if we are interactive
	 * or the process which created us was.
	 */
	if (intact || parintr == SIG_DFL)
		**av = '-';

	/*
	 * Save the remaining arguments in ARGV.
	 * Normally the system-supplied argument list is ok as
	 * a zero terminated value block.
	 * On some version 6 systems, it is -1 terminated and setting it
	 * to zero messes up "ps" so we change it to zero, copy
	 * the block of pointers, and put it back the way it was.
	 */
/*
	if (c == 0)
		set("argv", 0);
	else
 */
	if ((int) v[c] == -1) {
		/* ick */
		v[c] = 0, setq("argv", copyblk(v), &shvhed), v[c] = (char *) -1;
	} else
		setq("argv", v, &shvhed);

	/*
	 * Set up the prompt.
	 */
	if (prompt)
		set("prompt", uid == 0 ? "# " : "% ");

	/*
	 * If we are an interactive shell, then start fiddling
	 * with the signals; this is a tricky game.
	 */
	if (**av == '-') {
		setintr++;
		if (!quitit)		/* Wary! */
			signal(SIGQUIT, SIG_IGN);
		signal(SIGINT, SIG_IGN);
		signal(SIGTERM, SIG_IGN);
	}

	/*
	 * Set an exit here in case of an interrupt or error reading
	 * the shell start-up scripts.
	 */
	setexit();
	haderr = 0;		/* In case second time through */
	if (!fast && reenter == 0) {
		reenter++;
		if (loginsh)
			srccat("", "/etc/cshrc");
		/* Will have value("home") here because set fast if don't */
		srccat(value("home"), "/.cshrc");
		if (!fast && !arginp && !onelflg)
			dohash();
		if (loginsh)
#ifdef NOHELP
			srccat("", ".login");
#else
			srccat(value("home"), "/.login");
#endif
	}

	/*
	 * Now are ready for the -v and -x flags
	 */
	if (nverbose)
		setNS("verbose");
	if (nexececho)
		setNS("echo");

	/*
	 * All the rest of the world is inside this call.
	 * The argument to process indicates whether it should
	 * catch "error unwinds".  Thus if we are a interactive shell
	 * our call here will never return by being blown past on an error.
	 */
	process(setintr);

	/*
	 * Mop-up.
	 */
	if (loginsh) {
		printf("logout\n");
		close(SHIN);
		child++;
		goodbye();
	}
	exitstat();
}

importpath(cp)
char *cp;
{
	register int i = 0;
	register char *dp;
	register char **pv;
	int c;
	static char dot[2] = {'.', 0};

	for (dp = cp; *dp; dp++)
		if (*dp == ':')
			i++;
	/*
	 * i+2 where i is the number of colons in the path.
	 * There are i+1 directories in the path plus we need
	 * room for a zero terminator.
	 */
	pv = (char **) Calloc(i+2, sizeof (char **));		/* L005 */
	dp = cp;
	i = 0;
	for (;;) {
		if ((c = *dp) == ':' || c == 0) {
			*dp = 0;
			pv[i++] = savestr(*cp ? cp : dot);
			if (c) {
				cp = dp + 1;
				*dp = ':';
			} else
				break;
		}
		dp++;
	}
	pv[i] = 0;
	set1("path", pv, &shvhed);
	dohash();
}

/*
 * Source to the file which is the catenation of the argument names.
 */
srccat(cp, dp)
	char *cp, *dp;
{
	register char *ep = strspl(cp, dp);
	register int unit = dmove(open(ep, 0), -1);

	/* ioctl(unit, FIOCLEX, NULL); */
	xfree(ep);
	srcunit(unit, 0);
}

/*
 * Source to a unit.  If onlyown it must be our file or
 * we don't chance it.	This occurs on ".cshrc"s and the like.
 */
srcunit(unit, onlyown)
	register int unit;
	bool onlyown;
{
	/* We have to push down a lot of state here */
	/* All this could go into a structure */
	volatile int oSHIN = -1;				/* L006 L007 */
	int oldintty = intty;
	struct whyle *oldwhyl = whyles;
	char *ogointr = gointr, *oarginp = arginp;
	int oonelflg = onelflg;
#ifdef TELL
	bool otell = cantell;
#endif
	struct Bin saveB;

	/* The (few) real local variables */
	jmp_buf oldexit;
	volatile int reenter;					/* L006 L007 */
	register void (*oldint)();

	if (unit < 0)
		return;
	if (onlyown) {
		struct stat stb;

#ifdef V69
		if (fstat(unit, &stb) < 0 || (stb.st_uid != uid && stb.st_uid != (uid &~ 0377))) {
#else
		if (fstat(unit, &stb) < 0 || (stb.st_uid != uid && stb.st_gid != getgid())) {
#endif
			close(unit);
			return;
		}
	}

	/*
	 * There is a critical section here while we are pushing down the
	 * input stream since we have stuff in different structures.
	 * If we weren't careful an interrupt could corrupt SHIN's Bin
	 * structure and kill the shell.
	 *
	 * We could avoid the critical region by grouping all the stuff
	 * in a single structure and pointing at it to move it all at
	 * once.  This is less efficient globally on many variable references
	 * however.
	 */
	getexit(oldexit);
	reenter = 0;
	oldint = signal(SIGINT, SIG_IGN);
	setexit();
	reenter++;
	if (reenter == 1) {
		/* Setup the new values of the state stuff saved above */
		copy(&saveB, &B, sizeof saveB);
		fbuf = (char **) 0;
		fseekp = feobp = fblocks = 0;
		oSHIN = SHIN, SHIN = unit, arginp = 0, onelflg = 0;
		intty = isatty(SHIN), whyles = 0, gointr = 0;
		/*
		 * Now if we are allowing commands to be interrupted,
		 * we let ourselves be interrupted.
		 */
		signal(SIGINT, setintr ? pintr : oldint);
#ifdef TELL
		settell();
#endif
		process(0);		/* 0 -> blow away on errors */
	}
	signal(SIGINT, oldint);
	if (oSHIN >= 0) {
		register int i;

		/* We made it to the new state... free up its storage */
		/* This code could get run twice but xfree doesn't care */
		for (i = 0; i < fblocks; i++)
			xfree(fbuf[i]);
		xfree(fbuf);

		/* Reset input arena */
		copy(&B, &saveB, sizeof B);

		close(SHIN), SHIN = oSHIN;
		arginp = oarginp, onelflg = oonelflg;
		intty = oldintty, whyles = oldwhyl, gointr = ogointr;
#ifdef TELL
		cantell = otell;
#endif
	}

	resexit(oldexit);
	/*
	 * If process reset() (effectively an unwind) then
	 * we must also unwind.
	 */
	if (reenter >= 2)
		error(0);
}

goodbye()
{

	if (loginsh) {
		signal(SIGQUIT, SIG_IGN);
		signal(SIGINT, SIG_IGN);
		signal(SIGTERM, SIG_IGN);
		setintr = 0;		/* No interrupts after "logout" */
		if (adrof("home"))
			srccat(value("home"), "/.logout");
	}
	exitstat();
}

exitstat()
{

	/*
	 * Note that if STATUS is corrupted (i.e. getn bombs)
	 * then error will exit directly because we poke child here.
	 * Otherwise we might continue unwarrantedly (sic).
	 */
	child++;
	exit(getn(value("status")));
}

/*
 * Catch an interrupt, e.g. during lexical input.
 * If we are an interactive shell, we reset the interrupt catch
 * immediately.  In any case we drain the shell output,
 * and finally go through the normal error mechanism, which
 * gets a chance to make the shell go away.
 */
void
pintr(int unused)
{
	register char **v;

	if (setintr)
		signal(SIGINT, SIG_IGN);
	draino();

	/*
	 * If we have an active "onintr" then we search for the label.
	 * Note that if one does "onintr -" then we shan't be interruptible
	 * so we needn't worry about that here.
	 */
	if (gointr) {
		search(ZGOTO, 0, gointr);
		timflg = 0;
		if (v = pargv)
			pargv = 0, blkfree(v);
		if (v = gargv)
			gargv = 0, blkfree(v);
		reset();
	} else if (intty)
		printf("\n");		/* Some like this, others don't */
	error(0);
}

/*
 * Process is the main driving routine for the shell.
 * It runs all command processing, except for those within { ... }
 * in expressions (which is run by a routine evalav in sh.exp.c which
 * is a stripped down process), and `...` evaluation which is run
 * also by a subset of this code in sh.glob.c in the routine backeval.
 *
 * The code here is a little strange because part of it is interruptible
 * and hence freeing of structures appears to occur when none is necessary
 * if this is ignored.
 *
 * Note that if catch is not set then we will unwind on any error.
 * In an end-of-file occurs, we return.
 */
process(catch)
	bool catch;
{
	register char *cp;
	jmp_buf osetexit;
	struct command *t;

	getexit(osetexit);
	for (;;) {
		paraml.next = paraml.prev = &paraml;
		paraml.word = "";
		t = 0;
		setexit();
		justpr = 0;			/* A chance to execute */

		/*
		 * Interruptible during interactive reads
		 */
		if (setintr)
			signal(SIGINT, pintr);

		/*
		 * For the sake of reset()
		 */
		freelex(&paraml), freesyn(t), t = 0;

		if (haderr) {
			if (!catch) {
				/* unwind */
				doneinp = 0;
				resexit(osetexit);
				reset();
			}
			haderr = 0;
			/*
			 * Every error is eventually caught here or
			 * the shell dies.  It is at this
			 * point that we clean up any left-over open
			 * files, by closing all but a fixed number
			 * of pre-defined files.  Thus routines don't
			 * have to worry about leaving files open due
			 * to deeper errors... they will get closed here.
			 */
			closem();
			continue;
		}
		if (doneinp) {
			doneinp = 0;
			break;
		}
		if (intty) {
			mailchk();
			/*
			 * If we are at the end of the input buffer
			 * then we are going to read fresh stuff.
			 * Otherwise, we are rereading input and don't
			 * need or want to prompt.
			 */
			if (fseekp == feobp)
				if (!whyles)
					for (cp = value("prompt"); *cp; cp++)
						if (*cp == HIST)
							printf("%d", eventno + 1);
						else {
							if (*cp == '\\' && cp[1] == HIST)
								cp++;
							putchar(*cp | QUOTE);
						}
				else
					/*
					 * Prompt for forward reading loop
					 * body content.
					 */
					printf("? ");
			flush();
		}
		err = 0;

		/*
		 * Echo not only on VERBOSE, but also with history expansion.
		 * If there is a lexical error then we forego history echo.
		 */
		if (lex(&paraml) && !err && intty || adrof("verbose")) {
			haderr = 1;
			prlex(&paraml);
			haderr = 0;
		}

		/*
		 * The parser may lose space if interrupted.
		 */
		if (setintr)
			signal(SIGINT, SIG_IGN);

		/*
		 * Save input text on the history list if it
		 * is from the terminal at the top level and not
		 * in a loop.
		 */
		if (catch && intty && !whyles)
			savehist(&paraml);

		/*
		 * Print lexical error messages.
		 */
		if (err)
			error(err);

		/*
		 * If had a history command :p modifier then
		 * this is as far as we should go
		 */
		if (justpr)
			reset();

		alias(&paraml);

		/*
		 * Parse the words of the input into a parse tree.
		 */
		t = syntax(paraml.next, &paraml, 0);
		if (err)
			error(err);

		/*
		 * Execute the parse tree
		 */
		execute(t);

		/*
		 * Made it!
		 */
		freelex(&paraml), freesyn(t);
	}
	resexit(osetexit);
}

dosource(t)
	register char **t;
{
	register char *f;
	register int u;

	t++;
	f = globone(*t);
	u = dmove(open(f, 0), -1);
	xfree(f);
	if (u < 0)
		Perror(f);
	didfds = 0;
	srcunit(u, 0);
}

/*
 * Check for mail.
 * If we are a login shell, then we don't want to tell
 * about any mail file unless its been modified
 * after the time we started.
 * This prevents us from telling the user things he already
 * knows, since the login program insists on saying
 * "You have mail."
 */
mailchk()
{
	register struct varent *v;
	register char **vp;
	time_t t;
	int intvl, cnt;

	v = adrof("mail");
	if (v == 0)
		return;
	time(&t);
	vp = v->vec;
	cnt = blklen(vp);
	intvl = (cnt && number(*vp)) ? (--cnt, getn(*vp++)) : MAILINTVL;
	if (intvl < 1)
		intvl = 1;
	if (chktim + intvl > t)
		return;
	for (; *vp; vp++) {
		bool new;
		struct stat stb;

		if (stat(*vp, &stb) < 0)
			continue;
		/*
		 * We assume that a file has been read if the access time is
		 * greater than the mod time.
		 */
		if (stb.st_size == 0)
			continue;
		if (stb.st_atime > stb.st_mtime || (stb.st_atime < chktim && stb.st_mtime < chktim))
			continue;
		new = stb.st_mtime > time0;
		if (loginsh && !new)
			continue;
		if (cnt == 1)
			printf("You have %smail.\n", new ? "new " : "");
		else
			printf("%s in %s.\n", new ? "New mail" : "Mail", *vp);
	}
	chktim = t;
}

#include <pwd.h>
/*
 * Extract a home directory from the password file
 * The argument points to a buffer where the name of the
 * user whose home directory is sought is currently.
 * We write the home directory of the user back there.
 */
gethdir(home)
	char *home;
{
	register struct passwd *pp;

	pp = getpwnam(home);
	if (pp == 0)
		return (1);
	strcpy(home, pp->pw_dir);
	return (0);
}

/*
 * Move the initial descriptors to their eventual
 * resting places, closin all other units.
 */
initdesc()
{

	didcch = 0;			/* Havent closed for child */
	didfds = 0;			/* 0, 1, 2 aren't set up */
	SHIN = dcopy(0, FSHIN);
	SHOUT = dcopy(1, FSHOUT);
	SHDIAG = dcopy(2, FSHDIAG);
	OLDSTD = dcopy(SHIN, FOLDSTD);
	closem();
}

#ifndef V6
exit(i)
	int i;
{

	_exit(i);
}
#endif
